platform9/ui-components
# Fix useInterval bug

**Bug:**

This is an issue with `useInterval` hook and `PollingData` component. I noticed that when I refresh a page that has the `PollingData` component, the data doesn't automatically refresh after a certain interval. 

**How to reproduce:** 

1. Go to a page that uses the PollingData component. For example, the CAPI clusters list page in PMK.
2. Open the networks tab to view the network calls
3. Refresh the page
4. Notice how no API calls are made. We should be making calls every 30 seconds


**Why is not working?**

When we refresh the page, the loading prop from our actions goes from True to False

I noticed that in the `PollingData` component, when we pass the callback function to the `useInterval` hook, the `loading` variable evaluates to `true` and it stays as `true` even though the `loading` prop in `PollingData` component has become `false`

PollingData.tsx
```
  const [lastIntervalTs, setLastIntervalTs] = useState(new Date().valueOf())
  
   ....

  if (!pause) {
    useInterval(() => {
      if (!loading) {
        setLastIntervalTs(new Date().valueOf())
      }
    }, pollIntervalMs)

    const currentTs = new Date().valueOf()
    if (currentTs - lastFetchTs > refreshDuration) {
      reload()
    }
  }
```

I added some console.logs

```
  if (!pause) {
    console.log('Outside callback', loading)
    useInterval(() => {
      if (!loading) {
        console.log('Inside callback, loading)
        setLastIntervalTs(new Date().valueOf())
      }
    }, pollIntervalMs)
  }
```

This is the output

![Screenshot 2023-10-13 at 5 25 57 PM](https://github.com/platform9/ui-components/assets/23369276/7624bd21-18cd-4aa3-9857-1a9fc02de8d7)

Since `loading === true`, the `setLastIntervalTs(new Date().valueOf())` never gets called. Since the state doesn't get changed, the lines below don't get evaluated

PollingData.tsx
```
    const currentTs = new Date().valueOf()
    if (currentTs - lastFetchTs > refreshDuration) {
      reload()
    }
```

The reason the `loading` variable is always `true` in the callback is because how we pass in the savedCallbackRef in `useInterval`

```
const useInterval = (callback: ICallback, delay = 0): void => {
  const savedCallbackRef = useRef<ICallback>(callback)
  const intervalIdRef = useRef<NodeJS.Timer>()

  useEffect((): void => {
    savedCallbackRef.current = callback
  }, [callback])

  useEffect(() => {
    intervalIdRef.current && clearInterval(intervalIdRef.current)
    if (delay != null) {
      intervalIdRef.current = setInterval(savedCallbackRef.current, delay)   <--- this line
    }
    return () => intervalIdRef.current && clearInterval(intervalIdRef.current)
  }, [delay])
}
```

`savedCallbackRef.current` gets evaluated only once when we start the interval. That is why it has the old value of `loading`

`intervalIdRef.current = setInterval(savedCallbackRef.current, delay)`

should be 

`intervalIdRef.current = setInterval(() => { savedCallbackRef.current() }, delay)`
